contracts/mesh-provider/src/ibc.rs

```diff
@@ -111,26 +118,53 @@ pub fn ibc_packet_receive(
 
     let msg: ConsumerMsg = from_slice(&msg.packet.data)?;
     match msg {
-        ConsumerMsg::Rewards {} => receive_rewards(deps),
+        ConsumerMsg::Rewards {
+            validator,
+            total_funds,
+        } => receive_rewards(deps, env, validator, total_funds),
         ConsumerMsg::UpdateValidators { added, removed } => {
-            receive_update_validators(deps, added, removed)
+            receive_update_validators(deps, env, added, removed)
         }
     }
 }
 
-pub fn receive_rewards(_deps: DepsMut) -> Result<IbcReceiveResponse, ContractError> {
-    // TODO
-    unimplemented!();
+pub fn receive_rewards(
+    deps: DepsMut,
+    env: Env,
+    validator: String,
+    total_funds: Coin,
+) -> Result<IbcReceiveResponse, ContractError> {
+    // Update the rewards of this validator
+    // This will fail if we didn't add the validator before it, we cannot init the validator and calculate rewards in the same msg. (same block)
+    let res = VALIDATORS.update::<_, ContractError>(deps.storage, &validator, |val| {
+        let mut val = match val {
+            Some(val) => val,
+            None => return Err(ContractError::ValidatorRewardsCalculationWrong {}),
+        };
+
+        val.rewards.calc_rewards(total_funds.amount, val.shares_to_tokens(val.stake), env.block.height)?;
+
+        Ok(val)
+    });
+
+    // TODO: if calculation failed, we want to handle it as leftover funds? or send funds back to consumer and handle it there?
```

Comment:
good question. no idea.
maybe raise some issues for these TODOS 
(or one meta-issue with checkboxes linking to these TODOs)

Reply:
Will take care of this when this PR is merged.